### PR TITLE
1854765 literal transition properties take precedence

### PR DESCRIPTION
**Pull request by dlgroves**
- Given: A RIM for a resource with three transition properties A, B and C
- When: The resource is requested
- Then: Literal transition properties declared in the RIM must overwrite identically named request query parameters when links' href attributes are built

### CODE REVIEW

**Comment by ikarady** [comment]
The if-else at lines 176-180 seems unnecessary, can we replace it with "outQueryParams.putSingle(key, paramValue);"?

It seems that we're trying to make sure that URI parameters will always override query parameters in linkTemplate. If that is the case then can we just simply use replaceQueryParam() on linkTemplate when adding URI parameters to implement this functionality (instead of introducing a temporary multi value map object)?

The unit test doesn't seem to test this new functionality as the InteractionContext passed into LinkGenerator is always null, also URI- and query parameters are different so they will never override each other.

File: interaction-core/src/main/java/com/temenos/interaction/core/hypermedia/LinkGeneratorImpl.java [hunk 1]
```diff
@@ -166,17 +157,37 @@ private void configureLink(UriBuilder linkTemplate, Map<String, Object> transiti
...
+        
         if (uriParameters != null) {
             for (String key : uriParameters.keySet()) {
                 String value = uriParameters.get(key);
                 if (!targetResourcePath.contains("{" + key + "}")) {
                     String paramValue = HypermediaTemplateHelper.templateReplace(value, transitionProperties);
                     if (paramValue != null) {
-                        linkTemplate.queryParam(key, paramValue);
+                        if(outQueryParams.containsKey(key)){
+                            outQueryParams.putSingle(key, paramValue);
```
Comment (cmclopes) [inline interaction-core/src/main/java/com/temenos/interaction/core/hypermedia/LinkGeneratorImpl.java]: @ikarady This line is to replace all the params received by the initial state with the param definition in the transition, example:

`/resourceA?param1=A,param2=B,param1=C`

If we have a autotransition on resourceA to resourceB that sets` [ param1 = 'Z' ]`

I'm expecting to receive as result of transition (link):

`/resourceB?param1=Z,param2=B`

File: interaction-core/src/test/java/com/temenos/interaction/core/hypermedia/TestLinkGeneratorImpl.java [hunk 2]
```diff
@@ -130,6 +130,20 @@ public void testCreateLinkHrefUriParameterTokensReplaceQueryParametersSpecial()
...
     }
+    
+    @Test
+    public void testCreateLinkHrefUriParameterEmptyTokensReplaceQueryParameters() {
+        ResourceStateMachine engine = new ResourceStateMachine(mock(ResourceState.class), new BeanTransformer());
+        Map<String, String> uriParameters = new HashMap<String, String>();
+        uriParameters.put("param1", "");
+        Transition t = new Transition.Builder().source(mock(ResourceState.class)).target(mockTarget("/test")).uriParameters(uriParameters).build();
+        MultivaluedMap<String, String> queryParameters = new MultivaluedMapImpl<String>();
+        queryParameters.add("$param", "123");
```
Comment (cmclopes) [inline interaction-core/src/test/java/com/temenos/interaction/core/hypermedia/TestLinkGeneratorImpl.java]: @dlgroves and @ikarady You're right, this line is wrong, this should be: 

```
queryParameters.add("param1", "123");
queryParameters.add("param2", "A");
queryParameters.add("param1", "Z");
```

and the last assert should be:

`assertEquals("/baseuri/test?param1=,param2=A", links.iterator().next().getHref());`

Makes sense?

**Comment by dlgroves** [comment]
Request query parameters named identically to transition properties will be substituted with the value of the transition property declared in the RIM.

File: interaction-core/src/test/java/com/temenos/interaction/core/hypermedia/TestLinkGeneratorImpl.java (hunk 2 above)
Comment (dlgroves) [inline interaction-core/src/test/java/com/temenos/interaction/core/hypermedia/TestLinkGeneratorImpl.java]: The test is incorrect as I will only get back /baseuri/test?param1= if I pass in my query parameters directly to LinkGenerator.createLink(); the test will pass if my InteractionContext's outQueryParameters have been set with these query names/values, however.

Comment (dlgroves) [inline interaction-core/src/test/java/com/temenos/interaction/core/hypermedia/TestLinkGeneratorImpl.java]: I'll add some additional test cases around InteractionContext parameters.

**Pull request merged**